atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -510,57 +505,44 @@ private void refreshTokenRanges() {
-            }
-            CassandraClientPoolingContainer hostPool = currentPools.get(specifiedHost);
-
-            if (blacklist.contains(specifiedHost) || hostPool == null || shouldRetryOnDifferentHost) {
-                InetSocketAddress previousHostPool = hostPool == null ? specifiedHost : hostPool.getHost();
-                Optional<CassandraClientPoolingContainer> hostPoolCandidate
-                        = getRandomGoodHostForPredicate(address -> !triedHosts.contains(address));
-                hostPool = hostPoolCandidate.orElseGet(this::getRandomGoodHost);
-                log.warn("Randomly redirected a query intended for host {} to {}.",
-                        SafeArg.of("previousHost", CassandraLogHelper.host(previousHostPool)),
-                        SafeArg.of("randomHost", CassandraLogHelper.host(hostPool.getHost())));
+                        SafeArg.of("host", CassandraLogHelper.host(req.getPreferredHost())));
             }
+            CassandraClientPoolingContainer hostPool = getPreferredHostOrFallBack(req);
 
             try {
-                return runWithPooledResourceRecordingMetrics(hostPool, fn);
+                return runWithPooledResourceRecordingMetrics(hostPool, req.getFunction());
             } catch (Exception e) {
-                numTries++;
-                triedHosts.add(hostPool.getHost());
-                this.<K>handleException(numTries, hostPool.getHost(), e);
-                if (isRetriableWithBackoffException(e)) {
-                    // And value between -500 and +500ms to backoff to better spread load on failover
-                    int sleepDuration = numTries * 1000 + (ThreadLocalRandom.current().nextInt(1000) - 500);
-                    log.warn("Retrying a query, {}, with backoff of {}ms, intended for host {}.",
-                            UnsafeArg.of("queryString", fn.toString()),
-                            SafeArg.of("sleepDuration", sleepDuration),
-                            SafeArg.of("hostName", CassandraLogHelper.host(hostPool.getHost())));
-
-                    try {
-                        Thread.sleep(sleepDuration);
-                    } catch (InterruptedException i) {
-                        throw new RuntimeException(i);
-                    }
-                    if (numTries >= MAX_TRIES_SAME_HOST) {
-                        shouldRetryOnDifferentHost = true;
-                    }
-                } else if (isFastFailoverException(e)) {
-                    log.info("Retrying with fast failover a query intended for host {}.",
-                            SafeArg.of("hostName", CassandraLogHelper.host(hostPool.getHost())));
-                    shouldRetryOnDifferentHost = true;
-                }
+                handleException(req, e, hostPool.getHost());
             }
         }
     }
 
+    private <V, K extends Exception> void handleException(RetryableCassandraRequest<V, K> req, Exception ex,
```

Comment:
This method adds no value, and can be inlined.

Reply:
Done.